File: server/utils/jsonParser.js
const logger = { // Basic logger, replace with a proper one if needed
    info: (msg) => console.log(\`[JSONParser INFO] \${msg}\`),
    warn: (msg) => console.warn(\`[JSONParser WARN] \${msg}\`),
    error: (msg) => console.error(\`[JSONParser ERROR] \${msg}\`)
};

class JSONParseException extends Error {
    constructor(message) {
        super(message);
        this.name = "JSONParseException";
    }
}

const JSONParser = {
    extractJsonFromContent: (content) => {
        if (!content || typeof content !== 'string') {
            logger.warn('Invalid content provided for JSON extraction.');
            return null;
        }

        const strategies = [
            // Standard markdown code block
            /```json\s*([\s\S]*?)\s*```/g,
            // Code block without 'json' tag but with braces (more lenient)
            /```\s*(\{[\s\S]*?\})\s*```/g,
            // Attempt to find a valid JSON object directly if not in code blocks
            // This regex is simplified and might need refinement for complex cases.
            // It tries to match balanced braces.
            /(\{(?:[^{}]|\{[^{}]*\})*\})/g
        ];

        for (const regex of strategies) {
            let match;
            // Reset lastIndex for global regexes if they are reused (though here we create new ones)
            // regex.lastIndex = 0;
            while ((match = regex.exec(content)) !== null) {
                let jsonStr = match[1].trim();
                jsonStr = JSONParser._cleanJsonString(jsonStr);
                try {
                    const result = JSON.parse(jsonStr);
                    if (typeof result === 'object' && result !== null) {
                        logger.info('Successfully parsed JSON from content.');
                        return result;
                    }
                } catch (e) {
                    logger.warn(\`Attempted to parse: \${jsonStr.substring(0,100)}... but failed: \${e.message}\`);
                    // Continue to try other matches or strategies
                }
            }
        }

        logger.warn('Could not find valid JSON in content using standard strategies. Attempting repair...');
        return JSONParser._attemptJsonRepair(content);
    },

    _cleanJsonString: (jsonStr) => {
        // Remove BOM if present (though less common in JS contexts from APIs)
        if (jsonStr.startsWith('\ufeff')) {
            jsonStr = jsonStr.substring(1);
        }
        // Some LLMs might add comments, try to remove them (very basic)
        jsonStr = jsonStr.split('\n').filter(line => !line.trim().startsWith('//')).join('\n');

        // Remove trailing commas (common issue)
        jsonStr = jsonStr.replace(/,\s*([}\]])/g, '$1');

        return jsonStr.trim();
    },

    _attemptJsonRepair: (content) => {
        logger.info('Attempting JSON repair...');
        let jsonPart = content;
        let startBrace = jsonPart.indexOf('{');
        if (startBrace === -1) {
            logger.warn('No opening brace found for JSON repair.');
            return null;
        }
        jsonPart = jsonPart.substring(startBrace);

        let braceCount = 0;
        let validEnd = -1;

        for (let i = 0; i < jsonPart.length; i++) {
            if (jsonPart[i] === '{') {
                braceCount++;
            } else if (jsonPart[i] === '}') {
                braceCount--;
                if (braceCount === 0) {
                    validEnd = i + 1;
                    break;
                }
            }
        }

        if (validEnd > 0) {
            const repairedJson = jsonPart.substring(0, validEnd);
            try {
                const result = JSON.parse(repairedJson);
                logger.info('Successfully repaired and parsed JSON.');
                return result;
            } catch (e) {
                logger.error(\`JSON repair failed: \${e.message}. Repaired string: \${repairedJson.substring(0,100)}...\`);
            }
        } else {
            logger.warn('Could not determine a valid end for JSON repair.');
        }
        return null;
    }
};

module.exports = { JSONParser, JSONParseException };

File: novel_writer/core/json_parser.py
"""
JSON解析器模組 - 從 novel_writer.py 重構抽取
處理從文本內容中提取和解析JSON數據的邏輯
"""

import json
import re
from typing import Dict, Optional


class JSONParser:
    """JSON解析器 - 重構版"""
    
    @staticmethod
    def extract_json_from_content(content: str) -> Optional[Dict]:
        """從內容中提取JSON"""
        strategies = [
            (r'```json\s*(.*?)\s*```', re.DOTALL),
            (r'```\s*(\{.*?\})\s*```', re.DOTALL),
            (r'\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\}', re.DOTALL),
        ]
        
        for pattern, flags in strategies:
            matches = re.findall(pattern, content, flags)
            
            for match in matches:
                json_str = match.strip() if isinstance(match, str) else match
                json_str = JSONParser._clean_json_string(json_str)
                
                try:
                    result = json.loads(json_str)
                    if isinstance(result, dict) and result:
                        return result
                except json.JSONDecodeError:
                    continue
        
        return JSONParser._attempt_json_repair(content)
    
    @staticmethod
    def _clean_json_string(json_str: str) -> str:
        """清理JSON字符串"""
        json_str = json_str.lstrip('\ufeff').strip()
        
        start_brace = json_str.find('{')
        if start_brace != -1:
            json_str = json_str[start_brace:]
        
        end_brace = json_str.rfind('}')
        if end_brace != -1:
            json_str = json_str[:end_brace + 1]
        
        return json_str
    
    @staticmethod
    def _attempt_json_repair(content: str) -> Optional[Dict]:
        """嘗試修復損壞的JSON"""
        json_start = content.find('{')
        if json_start == -1:
            return None
        
        json_part = content[json_start:]
        brace_count = 0
        valid_end = -1
        
        for i, char in enumerate(json_part):
            if char == '{':
                brace_count += 1
            elif char == '}':
                brace_count -= 1
                if brace_count == 0:
                    valid_end = i + 1
                    break
        
        if valid_end > 0:
            repaired_json = json_part[:valid_end]
            try:
                return json.loads(repaired_json)
            except json.JSONDecodeError:
                pass
        
        return None